### PR TITLE
Bug calculate stay

### LINKED ISSUE #351: Bug with `calculate_stay`
There is an issue with `calculate_stay`. If a `sc_qtr` is provided it will always ignore the `end_date`.

For example, the below should both return 1.
```R
> calculate_stay("1920", as.Date("2020-01-01"), as.Date("2020-01-02"))
[1] 1
> calculate_stay("1920", as.Date("2020-01-01"), as.Date("2020-01-02"), "2019Q4")
[1] 91
```

This was an oversight, as these cases aren't tested for either.

Tagging @Jennit07 as you wrote the function, please assign yourself if you're working on it, otherwise I will probably come back to it soon!

### PR DESCRIPTION
**Pull request by Jennit07**
@Moohan spotted a bug with the `calculate stay` function (great spot) where this was ignoring the `end_date` if `sc_qtr` was supplied. I think i've fixed this now and tested on the example in issue #351. Also added some tests to capture this as well, so hopefully this is working as it should now. Closes #351

### CODE REVIEW

File: R/calculate_stay.R [hunk 1]
```diff
@@ -38,5 +51,8 @@ calculate_stay <- function(year, start_date, end_date, sc_qtr = NULL) {
       # if end date is missing set to end of quarter
       lubridate::time_length(lubridate::interval(start_date, qtr_end), unit = "days")
     )
+  } else {
+    # Normal calculation if sc_qtr is supplied and end_date is not missing
+    lubridate::time_length(lubridate::interval(start_date, end_date), unit = "days")
```
Comment (Moohan) [inline R/calculate_stay.R:56]: This bit of code will never be executed since the above if / if else are binary i.e. one of them will always be executed, either `sc_qtr` is missing or it's not.

**Review by Moohan — CHANGES_REQUESTED** [review]
Great, fast work!

Great that the tests are passing but I think the logic could be tidied up, as there's a bit of redundant code and the logic for the `sc_qtr` is a bit hard to follow now. How about:

```R
} else {
# Set Quarters
    qtr_end <- lubridate::add_with_rollback(lubridate::yq(sc_qtr), lubridate::period(6, "months"))
    next_qtr <- lubridate::add_with_rollback(lubridate::yq(sc_qtr), lubridate::period(9, "months"))

dummy_end_date <- dplyr::case_when(
     !is.na(end_date) ~ end_date,
     qtr_end >= start_date ~ qtr_end,
     qtr_end < start_date ~ next_qtr
)

lubridate::time_length(lubridate::interval(start_date, dummy_end_date), unit = "days")
}

**Comment by Moohan** [comment]
I think you also need to update the snapshot test. I think if you run `devtools::test()` `testthat::test_package()` or click `Test` it should give you a message on how to review and accept the updated snap.

File: tests/testthat/_snaps/calculate_stay.md [hunk 2]
```diff
@@ -4,13 +4,13 @@
...
-       2 2019-06-30 2019-08-31 <NA>                    NA
-       3 2019-01-01 2020-04-01 <NA>                    NA
-       4 2019-04-01 2020-07-01 <NA>                    NA
+       1 2019-03-31 2019-10-31 <NA>                   214
+       2 2019-06-30 2019-08-31 <NA>                    62
+       3 2019-01-01 2020-04-01 <NA>                   456
+       4 2019-04-01 2020-07-01 <NA>                   457
        5 2019-03-31 NA         <NA>                    NA
        6 2019-06-30 NA         <NA>                    NA
        7 2019-01-01 NA         <NA>                    NA
```
Comment (Moohan) [inline tests/testthat/_snaps/calculate_stay.md:16]: What's happening here? Shouldn't these rows have a stay? 367, 267, 456 from your other tests.

File: tests/testthat/test-calculate_stay.R [hunk 3]
```diff
@@ -117,6 +127,24 @@ test_that("Can calculate the correct stay for vectors of dates", {
...
+        "2019/10/31", "2019/08/31",
+        "2020/04/01", "2020/07/01"
+      )),
+      c("2019Q1", "2019Q2", "2019Q3", "2019Q4")
+    ),
+    c(214, 62, 456, 457)
+  )
 })
 
 test_that("calculate stay function works", {
```
Comment (Moohan) [inline tests/testthat/test-calculate_stay.R:150]: Could you replace the tibble here with `tribble` please, as it's much easier to read (and add to):
```R
test_tibble <- tibble::tribble(
  ~start_date, ~end_date,  ~sc_latest_submission, 
  2019-03-31,  2019-10-31, NA,                 
  2019-06-30,  2019-08-31, NA,                 
  2019-01-01,  2020-04-01, NA,                 
  2019-04-01,  2020-07-01, NA,                 
  2019-03-31,  NA,         NA,                 
  2019-06-30,  NA,         NA,                 
  2019-01-01,  NA,         NA,                 
  2019-04-01,  NA,         NA,                 
  2019-03-31,  NA,         "2019Q1",             
  2019-06-30,  NA,         "2019Q2",             
  2019-01-01,  NA,         "2019Q3",             
  2019-04-01,  NA,         "2019Q4",             
  2019-07-31,  NA,         "2019Q1",             
  2019-10-31,  NA,         "2019Q2",             
  2020-01-31,  NA,         "2019Q3",             
  2020-04-30,  NA,         "2019Q4",             
  2019-03-31,  2019-10-31, "2019Q1",             
  2019-06-30,  2019-08-31, "2019Q2",             
  2019-01-01,  2020-04-01, "2019Q3",             
  2019-04-01,  2020-07-01, "2019Q4"
)
```

Comment (Jennit07) [reply tests/testthat/test-calculate_stay.R]: I've split out the tribble for normal/sc calculation. Could you please check this again? I can't get it to work properly and the tests keep failing. However, i have fixed the other tests this was failing on here too

File: tests/testthat/test-calculate_stay.R [hunk 4]
```diff
@@ -235,22 +235,28 @@ test_that("Can calculate the correct stay for vectors of dates", {
...
-    ~start_date, ~end_date,
-    2019 - 03 - 31, 2019 - 10 - 31,
-    2019 - 06 - 30, 2019 - 08 - 31,
-    2019 - 01 - 01, 2020 - 04 - 01,
-    2019 - 04 - 01, 2020 - 07 - 01,
-    2019 - 03 - 31, NA,
-    2019 - 06 - 30, NA,
-    2019 - 01 - 01, NA,
-    2019 - 04 - 01, NA
-  )
```
Comment (Moohan) [inline tests/testthat/test-calculate_stay.R]: @Jennit07 The issue was with the tribble definition (so my fault!) The dates weren't being declared properly.

**Review by Moohan — APPROVED** [review]
